rustls/src/conn.rs

```diff
@@ -81,12 +81,12 @@ impl Connection {
     /// Derives key material from the agreed connection secrets.
     ///
     /// See [`ConnectionCommon::export_keying_material()`] for more information.
-    pub fn export_keying_material(
+    pub fn export_keying_material<T: AsMut<[u8]>>(
         &self,
-        output: &mut [u8],
+        output: T,
```

Comment:
Why can't we just stick to `&mut [u8]` as input and `&[u8]` as output?

Reply:
We could, but then this won't fix the underlying risk factor in most usecases, particularly when follow on crypto libraries want types other than &[u8] as input (like for example RustCrypto, which wants GenericArrays everywhere).

Reply:
How about the caller is allowed to move in an `[u8; N]` with `const N: usize`, and we'll yield the same array?

Reply:
I guess this API doesn't see a lot of use, so breaking it (in a way that the compiler will detect) doesn't seem like a big deal to me.

Reply:
@djc, thinking a bit more about it, we don't really need to move in the [u8;N]. The functions is a const generic in N either way. How do you feel about that.

Reply:
Not sure how you think that changes things?

Reply:
For large returns, most calling conventions use a caller-allocated buffer, and I think (but not 100% sure) its the same for large parameters, so under the hood with two [u8;N] (one passed in, one passed out) there are potentially two buffers needed at the caller, because it might not be able to reason about whether both are live at the same time, which makes the memcpy reduction somewhat tricky. With just the return buffer, this becomes simpler for the compiler and therefore hopefully more likely to optimize any memcpys out. Plus as an input parameter the [u8;N] doesn't really do anything, in the sense that any value it contains is pretty much ignored (or at least, it should be).

Reply:
Ahh, so you're suggesting we pass in a `const N: usize` parameter but no buffer, and then the function can yield the `[u8; N]` as a return value? Makes sense to me!

Reply:
Reworked it to that. Apologies this took a while, been a bit swamped with other work.